test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java

```diff
@@ -0,0 +1,375 @@
+
+        public Mocker(String[] responses) throws IOException {
+            this.ss = new ServerSocket(0);
+            this.responses = responses;
+            this.requests = new LinkedList<>();
+        }
+
+        public void stopMocker() {
+            close(ss, s, in, out);
+        }
+
+        public int getPort() {
+            return ss.getLocalPort();
+        }
+
+        public String baseURL() {
+            try {
+                return URIBuilder.newBuilder()
+                    .scheme("http")
+                    .loopback()
+                    .port(getPort())
+                    .build()
+                    .toString();
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private String readRequest() throws IOException {
+            String req = "";
+            while (!req.endsWith("\r\n\r\n")) {
+                int x = in.read();
+                if (x == -1) {
+                    s.close();
+                    s = ss.accept();
+                    in = s.getInputStream();
+                    out = s.getOutputStream();
+                }
+                req += (char)x;
+            }
+            return req;
+        }
+
+        public String getRequest(int i) {
+            return requests.get(i);
+        }
+
+        public void run() {
+            try {
+                int index=0;
+                s = ss.accept();
+                in = s.getInputStream();
+                out = s.getOutputStream();
+                while (index < responses.length) {
+                    requests.add(readRequest());
+                    out.write(responses[index++].getBytes(US_ASCII));
+                }
+            } catch (Exception e) {
+                //e.printStackTrace();
+            }
+        }
+    }
+
+    static class ProxySel extends ProxySelector {
+        final int port;
+
+        ProxySel(int port) {
+            this.port = port;
+        }
+        @Override
+        public List<Proxy> select(URI uri) {
+          return List.of(new Proxy(Proxy.Type.HTTP, new InetSocketAddress(
+              InetAddress.getLoopbackAddress(), port)));
+        }
+
+        @Override
+        public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {}
+
+    }
+
+    static class ProxyAuth extends Authenticator {
+        private volatile boolean called = false;
+
+        @Override
+        protected PasswordAuthentication getPasswordAuthentication() {
+            called = true;
+            return new PasswordAuthentication("proxyUser", "proxyPwd".toCharArray());
+        }
+
+        boolean wasCalled() {
+            return called;
+        }
+    }
+
+    static class ServerAuth extends Authenticator {
+        @Override
+        protected PasswordAuthentication getPasswordAuthentication() {
+            if (getRequestorType() != RequestorType.SERVER) {
+                // We only want to handle proxy authentication here
```

Comment:
I suspect this is a typo - I think it should have been "We only want to handle server ...."